File: c++/src/engine/collider.cc
#include "collider.h"

#include <cstdlib>

namespace Game {
    bool Collider::collides(const Rectangle & where) const {
        auto box = bbox();
        auto here = position();
        return
            std::abs(here.x + box.x - where.x) < (box.w + where.w) / 2 &&
            std::abs(here.y + box.y - where.y) < (box.h + where.h) / 2;
    }
}

File: js/src/engine/collider.js
'use strict';

import { Rectangle, Circle, Position } from './struct';

function Collider(bbox) {
  return function(Base = class{}) {
    return class extends Base {
      get position() {
        if(this.sprite) {
          return new Position(this.sprite.x, this.sprite.y);
        } else {
          throw `${this.constructor.name}'s position value cannot be inferred. Please implement '@override get position(): Position'`;
        }
      }
      // get the collision box for this object
      get bbox() {
        return bbox;
      }
      // actually for check a collision
      collides(where) {
        return this.bbox.constructor.intersects(this.bbox.constructor.shift(this.bbox, this.position), where);
      }
    };
  }
}

Object.defineProperty(Collider, Symbol.hasInstance, {
  value(instance) {
    return typeof instance.collides === 'function' &&
      instance.position instanceof Position &&
      (instance.bbox instanceof Rectangle || instance.bbox instanceof Circle);
  }
});

export { Collider };
export default Collider;
